src/rust/engine/process_execution/src/nailgun/mod.rs

```diff
@@ -86,7 +86,29 @@ fn construct_nailgun_client_request(
     is_nailgunnable,
   } = original_req;
   client_args.insert(0, client_main_class);
-  ExecuteProcessRequest {
+  // arg file is only materialized to the client workdir, but because we use nailMain java is running in the
+  // differnt  dir so we have to adjust the path to point to the correct spot, which is in the
+  // client workdir. This happens because the arg file is parsed before the absolute paths
+  // are computed by nailgun in nailMain
+  if let Some(maybe_arg_file) = client_args.last() {
```

Comment:
Rather than doing this (modifying client arguments like this is fragile), we should launch the nailgun server in the same directory that the client will run in.

To do that, we'd create a temporary directory to put the server's classpath in, but then we'd launch it with absolute paths for its classpath, and with its `cwd` == to the client directory. This would require rewriting the `-cp` (classpath) argument for the nailgun server to use absolute paths, but that is a simpler rewrite than this one.

Reply:
I agree this is fragile, but running the server and the client in the same directory would remove all the benefit of sharing code with the local command runner, and call for new mechanisms to materialize and snapshot a directory without removing it entirely. I assume the nailgun server wouldn't be happy if its CWD was removed, but it could be wrong.

What if instead we modify nailMain to look for arg files in the client directory? We already parse them out of the argv there. That would look like removing this block of code and adding some lines to the Zinc wrapper PR.

Reply:
> I assume the nailgun server wouldn't be happy if its CWD was removed, but it could be wrong.

True. But we wouldn't have to "remove" the client directory: just empty it between clients.

Reply:
We do expect and parse them in our zinc wrapper if they are passed. I'm proposing checking if they are in the nailMain cwd when parsing and using them there if present, and falling back to the server directory otherwise. We already update all paths in the arg file to the cwd when using nailMain, so it doesn't seem like a breach of contract to look for zinc args files in the cwd as well.

Reply:
... sure, but this code is not zinc specific. What we do in our zinc wrapper isn't relevant to our nailgun support "in general".

As mentioned elsewhere, some workarounds are ok until we have the ability not to require `nailMain` by spawning with appropriate cwds for the client/server but let's fully document the workarounds with TODOs linking to a ticket(s).